User profile migration documentation for default validations and stra… (#26732)
keycloak/keycloak: docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc

The usage of unversioned feature names, e.g. `docker`, in the `--features` list will allow for the most supported / latest feature version to be enabled for you. 
If you need more predictable behavior across releases, reference the particular version you want instead, e.g. `docker:v1`.

= User Profile Changes

> 

Author: It is plan to add all the other info related to user-profile migration (Like user-profile enabled by default, unmanaged attributes etc). In this PR, I've added only #25979 and #26634 .

== Default validations

Default user profile configuration comes with set of default validations for the basic predefined fields. Those validations were not present in previous versions when `declarative-user-profile` feature was disabled by default.

> If you have issues due the backwards compatibility, it is recommended that your users are updated according to latest validations. They are typically enforced to do so whenever they update their account

Reviewer: Isn't better to just remove any validation as they are business-specific?

IIRC, without user profile you would end up with the same validations as you were using the default user profile configuration, no?

The usage of unversioned feature names, e.g. `docker`, in the `--features` list will allow for the most supported / latest feature version to be enabled for you. 
If you need more predictable behavior across releases, reference the particular version you want instead, e.g. `docker:v1`.

= User Profile Changes

== Default validations

> Default user profile configuration comes with set of default validations for the basic predefined fields. Those validations were not present in previous versions when `declarative-user-profile` feature was disabled by default.

Reviewer 2: ```suggestion
Default user profile configuration comes with a set of default validations for the basic predefined fields. Those validations were not present in previous versions when the `declarative-user-profile` feature was disabled by default.
```

If you have issues due the backwards compatibility, you can change the default validators according to your needs. The default validators are as follows:

> * All of `username`, `email`, `firstName` and `lastName` attributes have maximum length 255 characters. These validations were indirectly present in previous versions as well

Reviewer 2: ```suggestion
* The`username`, `email`, `firstName` and `lastName` attributes have a maximum length of 255 characters. These validations were indirectly present in previous versions as well
```

> due there is DB constraint on the table `USER_ENTITY` for those fields with maximum length of 255 characters. However when using user storage providers, it might be possible before to use longer values.

Reviewer 2: ```suggestion
due there is a database constraint on the table `USER_ENTITY` for those fields with a maximum length of 255 characters. However when using user storage providers, it might be possible before to use longer values.
```

> * Attribute `username` has minimum length of 3 characters. Username has also `username-prohibited-characters` and `up-username-not-idn-homograph` validator by default, which were not present in previous versions.

Reviewer 2: ```suggestion
* The `username` attribute has a minimum length of three characters. Username has also `username-prohibited-characters` and `up-username-not-idn-homograph` validator by default, which were not present in previous versions.
```

> See link:{adminguide_link}#user-profile[Validation section of the User Profile Documentation] for the details about those. Note that username is not editable by default unless you have realm switch `Edit username enabled`.

Reviewer 2: ```suggestion
See the link:{adminguide_link}#user-profile[Validation section of the User Profile Documentation] for the details about those attributes. Note that username is not editable by default unless you have the realm switch `Edit username enabled`.
```

> This means that existing users with incorrect usernames should still work and they will not be enforced to update their usernames. But new users will be enforced to use correct usernames

Reviewer 2: ```suggestion
This change means that existing users with incorrect usernames should still work and they will not be enforced to update their usernames. But new users will be enforced to use correct usernames
```

> during their registration or creation by admin REST API.

Reviewer 2: ```suggestion
during their registration or creation by the admin REST API.
```

> * Attributes `firstName` and `lastName` have `person-name-prohibited-characters` validator on them, which were not present in previous versions.

Reviewer 2: ```suggestion
* Attributes `firstName` and `lastName` have the `person-name-prohibited-characters` validator on them, which were not present in previous versions.
```

> See link:{adminguide_link}#user-profile[Validation section of the User Profile Documentation] for the details about those. Note that both first

Reviewer 2: ```suggestion
See the link:{adminguide_link}#user-profile[Validation section of the User Profile Documentation] for the details about those attributes. Note that both first
```

> name and last name are editable by default, so users, who already have such incorrect first/last name from previous version will be forced to update them when updating their user profiles.

Reviewer 2: ```suggestion
name and last name are editable by default, so users, who already have such incorrect first/last name from a previous version will be forced to update them when updating their user profiles.
```

== User attribute names with strange characters

In previous versions, it was possible to create user with attribute names like `some:attribute` or `some/attribute`. User profile does not allow on purpose to create
attributes with such strange names in the user profile configuration. So it can be needed to configure `Unmanaged attributes` for your realm and enable unmanaged attributes for administrators
(ideally) or for end users (if really needed). Although it is strongly preferred to avoid using such attribute names.

== Verify Profile required action enabled by default

The required action `verify-profile` is enabled by default for new realms. But when you are migrating from previous version, your existing realms will have same state of this `verify-profile` action as before, which

> usually means disabled as it was disabled by default in previous versions. For the details about this required action, see link:{adminguide_link}#user-profile[User Profile Documentation].

Reviewer 2: ```suggestion
usually means disabled as it was disabled by default in previous versions. For the details about this required action, see the link:{adminguide_link}#user-profile[User Profile Documentation].
```
